rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,511 @@
+   tagged as “invalid”.
+   1. Expose this as an API first
+   2. (later) Notify administrators and allow them to export invalid objects
+      from the Kibana UI.
+4. When an invalid object is read, the Saved Objects repository will throw an
+   invalid object exception which should include a link to the documentation
+   to help administrators resolve migration bugs.
+5. Educate Kibana developers to no longer simply write back an unmigrated
+   document if an exception occurred. A migration function should either
+   successfully transform the object or throw. 
+
+## 4.3 Automatically retry failed migrations until they succeed
+
+> Achieves goals: (2.2)
+> Mitigates errors (3.3) and (3.4)
+
+External conditions such as failures from an unhealthy Elasticsearch cluster
+(3.3) can cause the migration to fail. The Kibana cluster should be able to
+recover automatically once these external conditions are resolved. There are
+two broad approaches to solving this problem based on whether or not
+migrations are idempotent: 
+
+| Idempotent migrations |Description                                                      |
+| --------------------- | --------------------------------------------------------------- |
+| Yes                   | 3.3.1 Idempotent migrations performed without coordination      |
+| No                    | 3.3.2 Single node migrations coordinated through a lease / lock |
+
+Because idempotent migrations don't require coordination, solution (3.3.1) is
+significantly less complex and it will never require manual intervention to
+retry. We, therefore, prefer this solution, even though it introduces new
+restrictions on migrations (4.3.1.1).
+
+## 4.3.1 Idempotent migrations performed without coordination
+
+The migration system can be said to be idempotent if the same results are
+produced whether the migration was run once or multiple times. This property
+should hold even if new (up to date) writes occur in between migration runs
+which introduces the following restrictions:
+
+### 4.3.1.1 Restrictions
+
+1. All document transforms need to be deterministic, that is a document
+   transform will always return the same result for the same set of inputs. 
+2. It should always be possible to construct the exact set of inputs required
+   for (1) at any point during the migration process (before, during, after). 
+
+### 4.3.1.2 Algorithm
+
+1. On startup, find the index to which the `.kibana` alias points,
+   `.kibana_n`, and check if any documents have an outdated `migrationVersion`
+   indicating that a migration needs to be performed. 
+2. Create `.kibana_n+1`, if it already exists, ignore the error and continue.
+3. Migrate documents by reading from `.kibana_n` and creating documents in
+   `.kibana_n+1`. If a document already exists, ignore.
+4. Move the `.kibana` index to point to `.kibana_n+1` (ignore if it has
+   already been moved). 
+
+Together with the limitations this algorithm ensures that migrations are
```

Comment:
I'm not sure how easy it is to create a thorough lint rule since it's not just side-effects like `fetch` or `fs` API's. A migration can be non-idempotent by introducing a random value too (like the spaces migration setting a document's id to `new uuid.v4()`)

I don't think we can create an automated test for this since it's hard to supply the right input document, but plugin's can test their migrations by running the migration on all the expected input shapes and asserting that the output matches a snapshot. If the same input generates the same output for all valid inputs it's safe.

Yeah that should be possible, we could either adopt a naming convention like `dashboard.migrations.ts` and apply linting based on the file name, or perform linting on the registered migration function itself (this won't lint code located in a function call). Neither would be a perfect guarantee.

Reply:
In java, we were asserting that migrations were idempotent by creating a large set of 'random' input objects using reflection and checking if the result was always the same when running the migration multiple times for each object. I wasn't really a fan of the approach but I have to admit this was working surprisingly well. This is probably not a realistic approach for JS though (even if `SavedObjectMigrationFn` now allows to type the input/output properties), so I guess delegating that testing to migration owners is probably our only option.

Reply:
> This is probably not a realistic approach for JS though

Yeah I think it would be hard, mostly b/c of the number of `string` fields we have that are really JSON. We could provide utilities for generating this data though?